### DISCUSSION TITLE
FMU in an iterative process in Python

### DISCUSSION THREAD

**Discussion by marcelotomim** [OpenModelica/OpenModelica#9879] [Q&A] [answered]
Dear All,

I am using omsimulator for co-simulating a FMU with a model I programmed in Python. There is a need to iterate both models for computing the results for a single time step. I used the command stepUntil with the same time stamp in order to calculate the same time step many times with updated inputs. When I compare results obtained with OMEdit, they drift apart as the simulation time moves forward, as one can see in the attached plots. What would be the best practice in the situation I am in?
I would really appreciate any feedback on this topic.
Thanks in advance for the attention.

[problema_4_3.pdf](https://github.com/OpenModelica/OpenModelica/files/10177025/problema_4_3.pdf)

**Comment by casella** [comment]
Try to reduce the step size by a factor 2 or a factor 10. If the drift is reduced, then it's just a matter of plain old truncation error 😃

**Reply by marcelotomim** [reply]
Thank you very much, Francesco, for your answer. I tried your suggestion before and, unfortunately, it did not work. The same drift remains. Would you have another idea? Is there another function I could use that would allow me repeat the step calculation without moving the time forward? It is important that the calculations remain synchronous.

**Comment by casella** [comment]
@arun3688, @lochel, maybe you can help here

**Reply by marcelotomim** [reply]
Hello Francesco.
I worked on my simulations again. And I got the drift under control changing the step size from 1e-3 to 1e-5, as you can seen in the attached plots.
But one question remains, how can I discard a step, during the iterative simulation? Is there a way to simply rollback to the previous step and perform the step calculation again?

[problema_4_3.pdf](https://github.com/OpenModelica/OpenModelica/files/10237462/problema_4_3.pdf)

**Comment by arun3688** [comment, marked as answer]
@marcelotomim To use the rollback feature in `OMSimulator` you have to use the select the  `VariableStepMaster`  solver which are 
`oms-mav` and `oms-mav2`,  I am attaching the report on how to use this rollback feature which contains the detailed description.
[Rollback_ResultAnalysis.pdf](https://github.com/OpenModelica/OpenModelica/files/10237536/Rollback_ResultAnalysis.pdf)

**Reply by marcelotomim** [reply]
Thank you @arun3688 for the information. I am using OMSimulator from Python, though. I run iteratively a FMU that is connected to a system modeled in Python. Would you have extra information on how I could use the rollback feature under such a circumstance? I would very much appreciate any advice on this.

**Comment by arun3688** [comment]
@marcelotomim You have to use  `fmi2GetFMUstate()` and `fmi2SetFMUstate`  to use the rollback feature something like below

```
# # for x in range(0, 100000000):
# #     # print(x)
# #     state = fmu.getFMUstate()
# #     fmu.setFMUState(state)
# #     fmu.freeFMUState(state)
```
Our OpenModelica fmus support the feature, but unfortunately we do not export the API in OMSimulator, as we do it internally

**Reply by marcelotomim** [reply]
Hi all,
Unfortunatelly, I was not able to sort this problem out. Is there a plan for implementing the save and get state im the FMU any time soon? Maybe I coud help?

**Reply by casella** [reply]
That would be really helpful. I'm not sure how hard it is, maybe you can briefly discuss it with @adrpo and @arun3688

**Reply by marcelotomim** [reply]
I assume this kind of implementation would require to change the omc compiler itself. Am I right? 
It would be great if I could help.

**Comment by arun3688** [comment]
@marcelotomim the fmu `getState` and `setState` is already implemented, but it is not exposed as an API in OMSimulator, I think you want those API's in your simulation loop ? And now we also have experimental prototype for serialization of FMU as described in the issue 
https://github.com/OpenModelica/OpenModelica/issues/9991

**Reply by marcelotomim** [reply]
@arun3688 I think that an access to getState and setState from OMSimulator would solve my problem. And, to be honest, I did not understand how the serialization prototype could help me.

**Comment by arun3688** [comment]
@marcelotomim I will soon expose the API in OMSimulator

**Reply by marcelotomim** [reply]
It would be simply perfect. Thanks in advance for the time and attention. 

One last thing, I would need to clarify. In order to be able to use the setState and getState, once exposed in the API, the FMU still need to implement these methods, right?

**Reply by arun3688** [reply]
@marcelotomim YEs the exporting FMU needs to implement these methods and our `OpenModelica` fmus have this support already

**Reply by marcelotomim** [reply]
Thanks for the information @arun3688 .
In case you need anything I am willing to help too.

**Comment by arun3688** [comment]
@marcelotomim I have now implemented the API for `getState, setState and freeState`  with this commit https://github.com/OpenModelica/OMSimulator/commit/4f5b1f767bb6817b320459b737dcf71b9ed5fb89, pull the latest changes from git and test it,  I have done some initial testing on the rollback feature and it looks good, Please refer the following example on how to use the API https://github.com/OpenModelica/OMSimulator/blob/master/testsuite/simulation/fmuState02.py

**Reply by marcelotomim** [reply]
This is great @arun3688 ! I am going to test it and keep you posted. Thanks for the your time and help.

**Reply by marcelotomim** [reply]
Hello @arun3688, 
I was able to compile the OMSimulator and tried to use the functions setState, getState and freeState. See enclosed the test I ran. Unfortunately, every time I call getState the script crashes. The FMU I am using, which was created with OMPython, is also inside the enclosed zip file.

Thanks again for your attention.

[rollback_test.zip](https://github.com/OpenModelica/OpenModelica/files/10679280/rollback_test.zip)

**Comment by arun3688** [comment]
@marcelotomim I do not have linux, So i cannot test the fmus, can you provide me the model to test the issue, from looking into the fmu provided by you i see that `canGetAndSetFMUstate="false" `  but it should be set to true, can you make the modification and test it otherwise please provide windows fmu to test

**Reply by marcelotomim** [reply]
How can I change this flag? The FMU is generated by OMPython. Is there another way I could generate the FMU?

**Comment by arun3688** [comment]
you can simply edit the fmu, open the archive and open the modeldescription.xml and change it to true, that is what i did, but unfortunately  the fmu has linux libraries so i could not test it. Can you provide the model so that i can generate the fmus myself

**Reply by marcelotomim** [reply]
I changed the flag and the script stopped crashing. The simulation runs smoothly until the system is perturbed, and the FMU starts returning NaN's. See enclosed the compressed file with the simulation script, the .mo file with the model that is compiled in the FMU by means of a second Python script based on OMPython.

[rollback_test_fmu.zip](https://github.com/OpenModelica/OpenModelica/files/10688563/rollback_test_fmu.zip)

Thanks again for the help.

**Comment by arun3688** [comment]
@marcelotomim does the fmu gets initialized correctly ?

**Reply by marcelotomim** [reply]
It does

**Reply by arun3688** [reply]
@marcelotomim I ran your script and i can see that it fails with non linear solver issue at different time steps, something like below,

```
LOG_NLS           | error   | residualFunc29: Iteration variable xloc[0] is nan.
assert            | debug   | residualFunc29 failed at time=0.104.
|                 | |       | For more information please use -lv LOG_NLS.
LOG_NLS           | error   | residualFunc29: Iteration variable xloc[0] is nan.
assert            | debug   | residualFunc29 failed at time=0.104.
|                 | |       | For more information please use -lv LOG_NLS.
LOG_NLS           | error   | residualFunc29: Iteration variable xloc[0] is nan.
assert            | debug   | residualFunc29 failed at time=0.104.
```
But i am not sure, the problem is related with `getStates `and `setStates`, To check that i have created a simple test case on the rollback feature see below

`fmuTestState.py`

```
from OMSimulator import OMSimulator
oms = OMSimulator()

oms.setCommandLineOption("--suppressPath=true")
oms.setTempDirectory("./model_03_py/")

oms.newModel("model")

oms.addSystem("model.root", oms.system_wc)

oms.addSubModel("model.root.system", "OmniPES.CoSimulation.Examples.Test_Classical_Machine.fmu")

oms.setStopTime("model", 5.0)

oms.setResultFile("model", "model03_res.mat")

oms.instantiate("model")
oms.initialize("model")

print("info: Initial Settings")

print("info:      model.root.system.data.Sbase     " , oms.getReal("model.root.system.data.Sbase")[0])
print("info:      model.root.system.gen1_data.MVAs " , oms.getReal("model.root.system.gen1_data.MVAs")[0])
print("info:      model.root.system.gen1_data.MVAb " , oms.getReal("model.root.system.gen1_data.MVAb")[0])
print("info:      model.root.system.gen1_data.H    " , oms.getReal("model.root.system.gen1_data.H")[0])
print("info:      model.root.system.gen1_data.D    " , oms.getReal("model.root.system.gen1_data.D")[0])

oms.stepUntil("model", 1.0)

Sbase = 2220.
oms.setReal("model.root.system.data.Sbase", Sbase)
oms.setReal("model.root.system.gen1_data.MVAs", Sbase)
oms.setReal("model.root.system.gen1_data.MVAb", Sbase)
oms.setReal("model.root.system.gen1_data.H", 3.5)
oms.setReal("model.root.system.gen1_data.D", 1.0)

print("info: New Start values")

print("info:      model.root.system.data.Sbase     " , oms.getReal("model.root.system.data.Sbase")[0])
print("info:      model.root.system.gen1_data.MVAs " , oms.getReal("model.root.system.gen1_data.MVAs")[0])
print("info:      model.root.system.gen1_data.MVAb " , oms.getReal("model.root.system.gen1_data.MVAb")[0])
print("info:      model.root.system.gen1_data.H    " , oms.getReal("model.root.system.gen1_data.H")[0])
print("info:      model.root.system.gen1_data.D    " , oms.getReal("model.root.system.gen1_data.D")[0])

oms.stepUntil("model", 2.0)

oms.getState("model.root.system")
oms.setState("model.root.system")
oms.freeState("model.root.system")

print("info: Rollback")
print("info:      model.root.system.data.Sbase     " , oms.getReal("model.root.system.data.Sbase")[0])
print("info:      model.root.system.gen1_data.MVAs " , oms.getReal("model.root.system.gen1_data.MVAs")[0])
print("info:      model.root.system.gen1_data.MVAb " , oms.getReal("model.root.system.gen1_data.MVAb")[0])
print("info:      model.root.system.gen1_data.H    " , oms.getReal("model.root.system.gen1_data.H")[0])
print("info:      model.root.system.gen1_data.D    " , oms.getReal("model.root.system.gen1_data.D")[0])
oms.terminate("model")
oms.delete("model")
```
### Result
```
arun@dev:~/OMSimulator/testsuite/rollback_test$ ../../install/linux/bin/OMSimulatorPython3 fmuState03.py
info:    Result file: model03_res.mat (bufferSize=1)
info: Initial Settings
info:      model.root.system.data.Sbase      100.0
info:      model.root.system.gen1_data.MVAs  100.0
info:      model.root.system.gen1_data.MVAb  100.0
info:      model.root.system.gen1_data.H     5.0
info:      model.root.system.gen1_data.D     0.0
info: New Start values
info:      model.root.system.data.Sbase      2220.0
info:      model.root.system.gen1_data.MVAs  2220.0
info:      model.root.system.gen1_data.MVAb  2220.0
info:      model.root.system.gen1_data.H     3.5
info:      model.root.system.gen1_data.D     1.0
info: Rollback
info:      model.root.system.data.Sbase      100.0
info:      model.root.system.gen1_data.MVAs  100.0
info:      model.root.system.gen1_data.MVAb  100.0
info:      model.root.system.gen1_data.H     5.0
info:      model.root.system.gen1_data.D     0.0

```
From the above results we can verify that rollback feature works, but may be some other issues with the iteration vars and start values where the non linear solver fails to converge or find a solution. Please test the above script to see whether rollback works

**Reply by marcelotomim** [reply]
Thanks @arun3688 
As I mentioned before, I got the getState, setState and freeState to work. It is weird, though, CVODE presents problems only when I used the saved state.

**Reply by arun3688** [reply]
@marcelotomim and also to generate the fmu you can directly use mos scripts and compile with omc instead of OMPython, An example mos script would be 

a.mos
```
setCommandLineOptions("-d=fmuExperimental,gendebugsymbols"); getErrorString();
installPackage(Modelica, "4.0.0+maint.om", exactMatch=true);
loadModel(Modelica, {"4.0"}); getErrorString();
loadFile("OmniPES.mo"); getErrorString();
buildModelFMU(OmniPES.CoSimulation.Examples.Test_Classical_Machine, fmuType="cs"); getErrorString();
```

and then simply use omc to execute the script

`>>> omc a.mos`

**Comment by marcelotomim** [comment]
Hello @arun3688. 
I started using the mos script above and it improved the FMU generation, compared to what I was doing before.

I would like to go back to the setState and getState functions, though.

I debugged my code line-by-line and I noticed that when I call the setState inside the iterative loop, the simulation time does not go back to the time I stored with the state when calling the getState function. Shouldn't the model go back in time with the previously stored state?

Thank again for you attention.

**Reply by marcelotomim** [reply]
Hi all,
It would very much appreciate if anybody could help with my previous question. Thanks in advance.

**Reply by arun3688** [reply]
@marcelotomim from the fmi2.0.2 specification `PAge 25, section 2.1.8`  it does not says anything about model go back in time to previously stored state, it says the following

```
The FMU has an internal state consisting of all values that are needed to continue a simulation. This 
internal state consists especially of the values of the continuous-time states, iteration variables, parameter 
values, input values, delay buffers, file identifiers, and FMU internal status information. With the functions
of this section, the internal FMU state can be copied and the pointer to this copy is returned to the 
environment. The FMU state copy can be set as actual FMU state, in order to continue the simulation from 
it.
```


see the fmi specification document
```
[FMI-Specification-2.0.2.pdf](https://github.com/OpenModelica/OpenModelica/files/11040167/FMI-Specification-2.0.2.pdf)
```

**Reply by marcelotomim** [reply]
Thank you @arun3688 . In this case, is there a way to stop the time from advancing during an iterative process, using a cosimulation FMU?

**Reply by arun3688** [reply]
@marcelotomim I am not sure about this ? @casella @AnHeuermann can you answer this

**Reply by AnHeuermann** [reply]
>  is there a way to stop the time from advancing during an iterative process, using a cosimulation FMU?

I don't think so. If you call fmi2DoStep the time will advance.

**Reply by arun3688** [reply]
@marcelotomim if You are using OMSimulator, then you can use the API, `oms_stepUntil`  for the particular time period

**Reply by marcelotomim** [reply]
What is the internal behavior of the stepUntil when I call it repeateadly with the same time?

**Comment by casella** [comment]
I don't have much hands-on experience here, but from what @arun3688 reported of the specification, with a bit of work you can save the state of an FMU, and then revert it later to it. But there's no explicit "back one step" function, you have to save and load this info yourself according to your need.

My 2 cts.

**Reply by marcelotomim** [reply]
From the tests I ran, I missed a function to inform the integrator that a step should be accepted without advancing the time during an iterative loop.
I was able to achieve this with a model exchange FMU and PyFMI.
